### ISSUE TITLE
At lot of users using gitbucket at the same time make it lock

### ISSUE THREAD

**Issue by darkweaver87** [gitbucket/gitbucket#1297]
## Issue
**Impacted version**:  4.4 and maybe before

**Deployment mode**:  standalone app

**Problem description**:
I'm using gitbucket as a standalone application on Ubuntu 14.04 with java 1.8.0 update 51. I have 204 accounts and 557 repositories.
When a lot of users are using gitbucket at the same time, gitbucket locks while trying to connect to its database (as far as I can understand, I'm not a java expert at all :-)).
It seems to be related to: https://github.com/brettwooldridge/HikariCP/issues/714.
I tried to migrate from H2 database to MySQL and it's a bit better but it is still locking with the following stakctrace:

```
 java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 30000ms.
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:196)
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:147)
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:83)
	at scala.slick.jdbc.JdbcBackend$DatabaseFactoryDef$$anon$3.createConnection(JdbcBackend.scala:47)
	at scala.slick.jdbc.JdbcBackend$BaseSession.conn$lzycompute(JdbcBackend.scala:397)
	at scala.slick.jdbc.JdbcBackend$BaseSession.conn(JdbcBackend.scala:397)
	at scala.slick.jdbc.JdbcBackend$BaseSession.withTransaction(JdbcBackend.scala:420)
	at scala.slick.backend.DatabaseComponent$DatabaseDef$$anonfun$withTransaction$1.apply(DatabaseComponent.scala:54)
	at scala.slick.backend.DatabaseComponent$DatabaseDef$$anonfun$withTransaction$1.apply(DatabaseComponent.scala:54)
	at scala.slick.backend.DatabaseComponent$DatabaseDef$class.withSession(DatabaseComponent.scala:34)
	at scala.slick.jdbc.JdbcBackend$DatabaseFactoryDef$$anon$3.withSession(JdbcBackend.scala:46)
	at scala.slick.backend.DatabaseComponent$DatabaseDef$class.withTransaction(DatabaseComponent.scala:54)
	at scala.slick.jdbc.JdbcBackend$DatabaseFactoryDef$$anon$3.withTransaction(JdbcBackend.scala:46)
	at gitbucket.core.servlet.TransactionFilter.doFilter(TransactionFilter.scala:28)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:581)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1174)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:511)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1106)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134)
	at org.eclipse.jetty.server.Server.handle(Server.java:524)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:319)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:253)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:273)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95)
	at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:93)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume(ExecuteProduceConsume.java:303)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:148)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:136)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:671)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:589)
	at java.lang.Thread.run(Thread.java:745)
```

I've been able to reproduce it by running multiple clones at the same time: 

```
for i in {1..20}; do git clone ssh://git.example.com:29418/user/repository.git repository_$i& done
```

Thanks.

Rémi

**Comment by McFoggy** [comment]
If we think it is a _normal_ behavior because of the lack of connections in the pool (compared to the number of incoming requests) perhaps we'll need to open a little bit the pool configuration. Doing that would allow intensive installations to modify the default pool configuration.

**Comment by takezoe** [comment]
A HikariCP's default value of max connection is `10`. I think it's enough for common use case.

Slick which is a database access framework being used in GitBucket shows better performance for PostgreSQL than MySQL. So it might be improved by migrating to PostgreSQL.

In addition, GitBucket has some slow queries. Maybe they degrade overall performance of your GitBucket instance. If we can find these slow queries, we can improve their performance.

Anyway, we can't configure HikariCP without modifying source code. I will update GitBucket to allow to configure HikariCP in `GITBUCKET_HOME/database.conf`.

**Comment by aadrian** [comment]
> Slick which is a database access framework being used in GitBucket shows better performance for PostgreSQL than MySQL. 

In all applications where we tried (given not with Slick or Scala) H2 performed way better (at least twice faster, but for many scenarios even more)  than PostgreSQL or MySQL. This also holds for scenarios with ~100 GB of data.

I'm wondering why isn't GitBucket using MirageSQL https://github.com/takezoe/mirage (since there's a Scala version) so that we can easily tune the SQLs for high performance.

**Comment by darkweaver87** [comment]
Thanks all ! I'll wait for the new release with HikariCP pool management then :-)

**Comment by McFoggy** [comment]
@takezoe I agree with you that we first undertand the root cause here. We are also perhaps keeping the connection too much in some endpoints.

**Comment by aadrian** [comment]
Maybe we could use some JMeter Load tests http://jmeter.apache.org/  to see the limits and be able to do finer tuning of params?

Not sure tough if Travis would allow load testing.

**Comment by takezoe** [comment]
@aadrian 
> I'm wondering why isn't GitBucket using MirageSQL https://github.com/takezoe/mirage (since there's a Scala version) so that we can easily tune the SQLs for high performance.

Slick has been considered a de-facto standard database access library in Scala world because Typesafe (currently Lightbend) is supported it and it was a part of the Typesafe reactive stack. We had to choose a library as major as possible for plug-in developers.

However Slick was removed from the Lightbend stack recently and development was moved to community based. Because of this, one of reasons staying Slick was lost in fact.
